test/ARCHITECTURE.md

```diff
@@ -0,0 +1,259 @@
+tests. The tests should automate how a normal user interacts with Cockpit.
+This requires a test machine which can add multiple disks or interfaces,
+reboot, interact with multiple machines on the same network and run potentially
+destructive test scenario's (e.g. installing/updating packages, formatting
+disks).
+
+For these reasons, Cockpit tests run inside a virtual machine (VM). The virtual
+machine uses Cockpit specific virtual machine images maintained and created in
+the [bots](https://github.com/cockpit-project/bots) repository. The images are
+usually based on a distribution's cloud image customized with:
+
+* A well-known password for the admin/root user
+* Test SSH keys for access
+* Test packages required to test Cockpit
+* A build chroot with Cockpit's build dependencies to build the to be tested
+  Cockpit source inside the virtual machine offline. This allows a developer on
+  Fedora to easily prepare a Debian test image without having to install Debian
+  build tools.
+* Disabling system services which interfere with testing
+
+To automate user actions on the web page, Cockpit is tested in a browser
+controlled using the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
+(CDP) which is supported by Firefox and Chromium based browsers.
+
+The Python test framework in [test/common](./common) is responsible for setting up the
+test environment, running tests and reporting of the test output.
+
+Diagram of the interaction of Browser/Machine/CDP/Test Framework.
+```mermaid
+graph TD;
+    id[Test Framework] <-->|CDP| Browser;
+    Browser <-->|HTTPS/WebSocket| Cockpit;
+    id1[Virtual Machine] <-->|HTTP/DBus/OS API's| Cockpit;
+    id[Test Framework] <-->|SSH| id1[Virtual Machine];
+```
+
+## Integration Tests
+
+Cockpit's tests can be run via three different entry points:
+
+* `test/verify/check-$page` - run a single or multiple unit test(s)
+* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties, parallelism)
+* `test/run` - CI entry point
+
+We will start with how a single integration test is run and then explore the test
+scheduler and CI setup.
+
+### Test runtime
+
+The base of a Cockpit integration test looks as following:
+
+```python
+class TestApps(testlib.MachineCase):
+    def testBasic(self):
+        self.machine.execute("rm /etc/machine-id")
+        self.login_and_go("/system")
+        self.browser.wait_text("#system_machine_id", "")
+
+if __name__ == '__main__':
+    testlib.test_main()
+```
+
+In Cockpit there are two types of tests, 'destructive" and "nondestructive".
+tests. Destructive tests do something to the virtual machine which makes it
+unable to run another test afterwards or requires another virtual machine for
+testing. The test above is a `destructive` test which is the default, a non
+`destructive` test makes sure any destructive action is restored after the test
```

Comment:
non `destructive` is weird -- first the conjunction with "non", and also because this isn't actually a valid decorator. non-destructive or `@nondestructive`